### DISCUSSION TITLE
"문제를 몇 초, 몇 분만에 풀었는가" feature 생성

### DISCUSSION THREAD

**Discussion by MoonJaeHoon** [bcaitech1/p4-dkt-freshtomato#31] [Q&A] [answered]
**문제를 몇 초, 몇 분만에 풀었는가** 라는 변수가 유의미할 것이기 때문에 생성해주고 싶었는데
groupby 메소드를 사용했더니 예외가 발생해서 어떻게 처리해야할지 의견 구합니다~!

> 코드 구현상의 문제점 : 해당 testid에 대하여 여러번 푼 적이 있는 사람에 대해서는 .groupby(['userid','testid']) 메소드를 사용해서 diff를 구할 수가 없다..

> 무작정 .groupby(['userid','testid']) 메소드를 사용하여 구했다가는 문제 풀이 간격이 `300일`인 경우도 발생하는 오류를 범하게 된다. 

> 그럼 중복되는 시험들에 예외처리까지 완벽하게 하기 위해서는 무조건 for문을 써야 하는걸까... (전처리시간 암담하당..)

아래는 groupby를 통해 구현해봤을 때의 예시이다.
-  value_counts()가 나타내는 index는 문제풀이 간격 일(day)이다)
- 보다시피 오류가 존재 : 1일~304일까지도 있다..

```python
.
.
.
one_problem_time_in_test = new_df.groupby(['userID','testId']).apply(lambda x: x['Timestamp'].diff())
one_problem_time_in_test.apply(lambda x: x.days if np.isnan(x.days) else int(x.days)).value_counts()
```
```
0.0      1893738
1.0          282
2.0          223
6.0          190
7.0          186
          ...   
217.0          1
304.0          1
232.0          1
266.0          1
239.0          1
```

**Comment by MignonDeveloper** [comment]
혹시 해당 feature를 ML에 쓸 수 있도록 user당 하나씩 만드는거야 아니면 Train Data의 Row당 하나씩 다 만들어주려고 하는거야?

**Reply by MoonJaeHoon** [reply]
모든 각 row마다 만들어주려고 하는 중이야!

`어떤 user가 하나의 test 안에서 문제를 풀때 각 하나의 문제당 풀이시간이 얼마나 걸렸는가?`
모든 row마다 이 부분을 feature로 주고 싶은 거지

**Comment by MignonDeveloper** [comment, marked as answer]
형이 말한대로 for문으로 밖에 생각이 안나서 해봤는데, 대략 10분정도 걸릴 것 같아...
```python
time_diff = pd.DataFrame(columns=["time_diff_in_one_test"])
copy_df = train_df.copy()
copy_df = copy_df[["testId", "Timestamp"]]
cur_test_id = copy_df.loc[0, "testId"]

for i in range(2266586 - 1):
    next_test_id = copy_df.loc[i+1, "testId"]
    if cur_test_id != next_test_id:
        cur_test_id = next_test_id
        time_diff.loc[i] = 0
    
    else:
        time_diff.loc[i] = (copy_df.loc[i+1, "Timestamp"] - copy_df.loc[i, "Timestamp"]).seconds

time_diff.loc[2266586 -1] = 0
time_diff.head(20)
```

**Reply by MignonDeveloper** [reply]
아니면, 사용자가 시험지를 풀때마다 unique한 label을 붙여주는 것도 되겠다!
user 별로 푸는 첫번째 시험지는 0, 두번째 시험지는 1, ... 이거를 따로 column을 추가해서 해보는건 어때?

이게 timestamp가 문제를 풀기 시작한 시점이니까 시험지별 마지막 문제는 얼마나 걸렸는지 시간을 측정할 수 가 없어서 우선 시험지별 마지막 문제는 0으로 하긴 했어 나는 ㅜ

**Reply by MignonDeveloper** [reply]
한가지 더는 나 같은 경우에 예전에 feature engineering할 때 진짜 너무 오래걸려서 그냥 한번 feature engineering한 Dataframe을 새로운 csv로 저장을 해버리고, 나중에 다시 학습할 때 그 친구를 불러와서 어떤 column을 사용할지 선택해주는 process를 중간에 추가했어!!

**Comment by arabae** [comment]
나도 user별 시험지 정답률 구하려고 for문으로 작성했는데 생각보다 오래 안걸렸어! <s>체감 못할정도였거든</s> 더 처리가 필요해서 오래거리는건가? 
➕ 시간도 측정해서 다시 알려줄께~

**Reply by MignonDeveloper** [reply]
누나가 사용한 for문 혹시 공유해줄 수 있을까? 거의 비슷한 작업인 것 같은데 나는 저거 돌리는데 10분이나 걸리더라고,, ㅜㅜ

**Reply by arabae** [reply]
나는 feature_engineering에 이렇게 작성해뒀어!
  
```python
def __feature_engineering(self, df):
        # TODO

        # --- user별 Test의 정답률 feature 추가
        total_count, answer_count = 1, 0
        pr_user_id, pr_test_id, pr_answer_code = 0, 'A060000001', 1
        user_test_correct_ratio = []

        for user_id, test_id, answer_code in zip(df['userID'], df['testId'], df['answerCode']):
            if (user_id != pr_user_id) or (test_id != pr_test_id): # 다른 user 시작 or 다른 시험지 시작
                total_count, answer_count = 1, 0
                pr_user_id = user_id
                pr_test_id = test_id
            if answer_code == 1:
                answer_count += 1
            user_test_correct_ratio.append(answer_count/total_count)
            total_count += 1
        
        user_test_correct_ratio = pd.DataFrame(user_test_correct_ratio, columns=['correctRate'])
        df = df.join(user_test_correct_ratio)

        return df
```

**Reply by arabae** [reply]
6.04388165473938초 걸렸습니당

**Reply by arabae** [reply]
엇 근데 이거 생각해보니까 나도 추가해보고싶었던 feature다!!
